src/Compression/CompressionCodecGCD.cpp

```diff
@@ -0,0 +1,273 @@
+#include <Compression/ICompressionCodec.h>
+#include <Compression/CompressionInfo.h>
+#include <Compression/CompressionFactory.h>
+#include <base/unaligned.h>
+#include <Parsers/IAST.h>
+#include <Parsers/ASTLiteral.h>
+#include <Parsers/ASTFunction.h>
+#include <IO/WriteHelpers.h>
+#include "Common/Exception.h"
+#include "base/Decimal_fwd.h"
+#include "base/types.h"
+#include "config.h"
+
+#include <cstdint>
+#include <numeric>
+
+namespace DB
+{
+
+template <typename T>
+T gcd_func(T a, T b)
+{
+    while (b != 0)
+    {
+        T c = a % b;
+        a = b;
+        b = c;
+    }
+    if (a == T(-1))
+    {
+        return -a;
+    }
+    return a;
+}
+
+class CompressionCodecGCD : public ICompressionCodec
+{
+public:
+    explicit CompressionCodecGCD(UInt8 gcd_bytes_size_);
+
+    uint8_t getMethodByte() const override;
+
+    void updateHash(SipHash & hash) const override;
+
+protected:
+    UInt32 doCompressData(const char * source, UInt32 source_size, char * dest) const override;
+    void doDecompressData(const char * source, UInt32 source_size, char * dest, UInt32 uncompressed_size) const override;
+    UInt32 getMaxCompressedDataSize(UInt32 uncompressed_size) const override;
+
+    bool isCompression() const override { return false; }
+    bool isGenericCompression() const override { return false; }
+
+private:
+    const UInt8 gcd_bytes_size;
+};
+
+
+namespace ErrorCodes
+{
+    extern const int CANNOT_COMPRESS;
+    extern const int CANNOT_DECOMPRESS;
+    extern const int BAD_ARGUMENTS;
+    extern const int ILLEGAL_SYNTAX_FOR_CODEC_TYPE;
+}
+
+UInt32 CompressionCodecGCD::getMaxCompressedDataSize(UInt32 uncompressed_size) const
+{
+    return uncompressed_size + 2;
+}
+
+CompressionCodecGCD::CompressionCodecGCD(UInt8 gcd_bytes_size_)
+    : gcd_bytes_size(gcd_bytes_size_)
+{
+    setCodecDescription("GCD", {std::make_shared<ASTLiteral>(static_cast<UInt64>(gcd_bytes_size))});
+}
+
+uint8_t CompressionCodecGCD::getMethodByte() const
+{
+    return static_cast<uint8_t>(CompressionMethodByte::GCD);
+}
+
+void CompressionCodecGCD::updateHash(SipHash & hash) const
+{
+    getCodecDesc()->updateTreeHash(hash);
+}
+
+namespace
+{
+
+template <typename T>
+void compressDataForType(const char * source, UInt32 source_size, char * dest)
+{
+    if (source_size % sizeof(T) != 0)
+        throw Exception(ErrorCodes::CANNOT_COMPRESS, "Cannot GCD compress, data size {}  is not aligned to {}", source_size, sizeof(T));
+
+    const char * const source_end = source + source_size;
+
+    T gcd{};
+    const auto * cur_source = source;
+    while (cur_source < source_end)
+    {
+        if (cur_source == source)
+        {
+            gcd = unalignedLoad<T>(cur_source);
+        }
+        else
+        {
+            gcd = gcd_func<T>(gcd, unalignedLoad<T>(cur_source));
+        }
+        if (gcd == T(1))
+        {
+            break;
+        }
+    }
+
+    unalignedStore<T>(dest, gcd);
```

Comment:
it should be extremely unlikely that the whole block of numbers has gcd>1. wdyt about breaking initial array into subarrays with gcd > 1 and encode them independently.
e.g. [2, 4, 8, 3, 9, 27] => [2 (gcd), 2, 4, 3 (gcd), 3, 9]. will require slight change in the algorithm to identify subarrays borders.

Reply:
This is likely with DateTime64 and Decimal columns (the GCD will be a power of ten).